src/core/ProjectVersion.cpp

```diff
@@ -27,123 +27,105 @@
+	// optional identifiers prefaced by a '-'. Both parts affect precedence
+	QString obligatorySegment = metadataStripped.section('-', 0, 0);
+	QString prereleaseSegment = metadataStripped.section('-', 1);
+
+	// The obligatory segment consists of three identifiers: MAJOR.MINOR.PATCH
+	QStringList mainVersion = obligatorySegment.split(".");
+	// HACK: Pad invalid versions in order to prevent crashes
+	while (mainVersion.size() < 3){ mainVersion.append("0"); }
+	m_major = mainVersion.at(0).toInt();
+	m_minor = mainVersion.at(1).toInt();
+	m_patch = mainVersion.at(2).toInt();
+
+	// Any # of optional pre-release identifiers may follow, separated by '.'s
+	if (!prereleaseSegment.isEmpty()){ m_labels = prereleaseSegment.split("."); }
+
+	// HACK: Handle old (1.2.2 and earlier), non-standard versions of the form
+	// MAJOR.MINOR.PATCH.COMMITS, used for non-release builds from source.
+	if (mainVersion.size() >= 4 && m_major <= 1 && m_minor <= 2 && m_patch <= 2){
+		// Drop the standard version identifiers. erase(a, b) removes [a,b)
+		mainVersion.erase(mainVersion.begin(), mainVersion.begin() + 3);
+		// Prepend the remaining identifiers as prerelease versions
+		m_labels = mainVersion + m_labels;
+		// Bump the patch version. x.y.z-a < x.y.z, but we want x.y.z.a > x.y.z
+		m_patch += 1;
+	}
 }
 
 
 
 
-ProjectVersion::ProjectVersion(const char* version, CompareType c) :
-	m_version(QString(version)),
-	m_major(parseMajor(m_version)),
-	m_minor(parseMinor(m_version)),
-	m_release(parseRelease(m_version)),
-	m_stage(parseStage(m_version)),
-	m_build(parseBuild(m_version)),
-	m_compareType(c)
+ProjectVersion::ProjectVersion(const char* version, CompareType c) : ProjectVersion(QString(version), c)
 {
 }
 
 
 
 
+//! @param c Determines the number of identifiers to check when comparing
 int ProjectVersion::compare(const ProjectVersion & a, const ProjectVersion & b, CompareType c)
 {
-	if(a.getMajor() != b.getMajor())
-	{
-		return a.getMajor() - b.getMajor();
-	}
-	if(c == Major)
-	{
-		return 0;
-	}
-
-	if(a.getMinor() != b.getMinor())
-	{
-		return a.getMinor() - b.getMinor();
-	}
-	if(c == Minor)
-	{
-		return 0;
+	// How many identifiers to compare before we consider the versions equal
+	const int limit = static_cast<int>(c);
+
+	// Use the value of limit to zero out identifiers we don't care about
+	int aMaj = 0, bMaj = 0, aMin = 0, bMin = 0, aPat = 0, bPat = 0;
+	if (limit >= 1){ aMaj = a.getMajor(); bMaj = b.getMajor(); }
+	if (limit >= 2){ aMin = a.getMinor(); bMin = b.getMinor(); }
+	if (limit >= 3){ aPat = a.getPatch(); bPat = b.getPatch(); }
+
+	// Then we can compare as if we care about every identifier
+	if(aMaj != bMaj){ return aMaj - bMaj; }
+	if(aMin != bMin){ return aMin - bMin; }
+	if(aPat != bPat){ return aPat - bPat; }
+
+	// Decide how many optional identifiers we care about
+	int numLabels = qMax(0, limit - 3);
+	auto aLabels = a.getLabels().mid(0, numLabels);
+	auto bLabels = b.getLabels().mid(0, numLabels);
```

Comment:
`QList::mid` allocates a new list - this could be avoided by taking `numLabels` into account later on instead, where the number of labels to compare is calculated.

Reply:
Notes to self just in case I change this, this line would need adjustment:
```
// If everything else matches, the version with more labels is bigger
return aLabels.size() - bLabels.size();
```

Reply:
I skipped this change because I felt that it convoluted the logic a bit. If it turns out that version comparisons are slowing things down significantly there are definitely optimizations to be had!